lnst/RecipeCommon/endpoints.py

```diff
@@ -0,0 +1,87 @@
+from __future__ import annotations
+from dataclasses import dataclass
+from typing import Generic, Iterator, Protocol, TypeVar
+
+from lnst.Common.IpAddress import BaseIpAddress
+from lnst.Controller.Namespace import Namespace
+from lnst.Devices.Device import Device
+
+
+class EndpointLike(Protocol):
+    @property
+    def host(self) -> Namespace:
+        ...
+
+    @property
+    def device_name(self) -> str:
+        ...
+
+
+TIpAddress_co = TypeVar("TIpAddress_co", bound=BaseIpAddress, covariant=True)
+
+
+class IPEndpointLike(EndpointLike, Protocol, Generic[TIpAddress_co]):
+    @property
+    def address(self) -> TIpAddress_co:
+        ...
+
+
+@dataclass
+class Endpoint:
+    device: Device
+
+    @property
+    def host(self) -> Namespace:
+        return self.device.netns
+
+    @property
+    def device_name(self) -> str:
+        return self.device.name
+
+
+@dataclass
+class IPEndpoint(Endpoint, Generic[TIpAddress_co]):
+    address: TIpAddress_co
+
+
+@dataclass
+class IPPortEndpoint(Generic[TIpAddress_co], IPEndpoint[TIpAddress_co]):
+    port: int
+
+
+TEndpoint = TypeVar("TEndpoint", bound=EndpointLike, covariant=True)
+TIpAddress = TypeVar("TIpAddress", bound=BaseIpAddress)
+
+
+@dataclass
+class EndpointPair(Generic[TEndpoint]):
+    first: TEndpoint
+    second: TEndpoint
+
+    def __iter__(self) -> Iterator[TEndpoint]:
+        yield self.first
+        yield self.second
+
+    @classmethod
+    def from_devices(cls, dev1: Device, dev2: Device) -> EndpointPair[Endpoint]:
+        return EndpointPair(
+            Endpoint(dev1),
+            Endpoint(dev2),
+        )
+
+    @classmethod
+    def from_devices_with_ips(
+        cls, dev1: Device, dev2: Device, ip_type: type[TIpAddress]
+    ) -> EndpointPair[IPEndpoint[TIpAddress]]:
+        selected_ips = []
+        for dev in [dev1, dev2]:
+            dev_ips = [ip for ip in dev.lnst_ips if isinstance(ip, ip_type)]
```

Comment:
not sure if this method should have filtering capability. we're already going to be applying the filtering where we use the results of this method.

IMO it's ok to have this method just always generate all of the device-ip endpoint configurations, and let the filtering be done by the user of the result afterwards.

Reply:
Changed the name to `all_with_ips` and made it yield all ip endpoint pairs. Let me know if you're happy with how I implemented it.

Reply:
Not relevant anymore.